teps/0055-migrate-podlatency-to-taskprelaunchlatency.md

```diff
@@ -0,0 +1,121 @@
+---
+status: proposed
+title: Migrate podLatency to taskPrelaunchLatency
+creation-date: '2021-01-06'
+last-updated: '2021-03-05'
+authors:
+- '@yaoxiaoqi'
+---
+
+# TEP-0055: Migrate `podLatency` to `taskPrelaunchLatency`
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+- [Alternatives](#alternatives)
+  - [Keeping the metrics <code>podLatency</code>](#keeping-the-metrics-)
+  - [Using a map to store the pod <code>startRunning</code> time](#using-a-map-to-store-the-pod--time)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+This TEP proposes to discard the metrics `podLatency` and add another metrics to
+report the pod creation -> pod running latency.
+
+## Motivation
+
+Currently, `podLatency` calculates the duration between pod creation time and
+pod scheduled time. The `podLatency` value is 0 at 99% of the time because
+Kubernetes timestamps are only at the resolution of a second, so if something
+takes less than a second it looks like it took 0 seconds. At the rest 1% of
+time, the metrics value is 1 or 2, which is not very worth to report. Therefore,
+this TEP proposes to discard the metrics `podLatency` and add another metrics to
+report the pod creation -> pod running latency. It can indicate how long a
+TaskRun pod takes to get ready for running.
```

Comment:
Is this meant for the first step in a `TaskRun` only or for each step? 

I guess this is for the first step only, so is the intention is to know how much time it takes to run the init-containers added by Tekton + download the image for the first step?

These are the events from a three-steps taskrun.
```
Events:
  Type    Reason     Age   From               Message
  ----    ------     ----  ----               -------
  Normal  Pulled     33s   kubelet            Container image "gcr.io/distroless/base@sha256:92720b2305d7315b5426aec19f8651e9e04222991f877cae71f40b3141d2f07e" already present on machine
  Normal  Created    33s   kubelet            Created container place-scripts
  Normal  Scheduled  33s   default-scheduler  Successfully assigned default/zero2cd-run-g75xw-build-n6gcc-pod-4jp2g to tekton-worker2
  Normal  Pulled     32s   kubelet            Container image "gcr.io/distroless/base@sha256:92720b2305d7315b5426aec19f8651e9e04222991f877cae71f40b3141d2f07e" already present on machine
  Normal  Created    32s   kubelet            Created container working-dir-initializer
  Normal  Started    32s   kubelet            Started container place-scripts
  Normal  Started    31s   kubelet            Started container working-dir-initializer
  Normal  Pulled     31s   kubelet            Container image "gcr.io/tekton-releases/github.com/tektoncd/pipeline/cmd/entrypoint:v0.23.0@sha256:5489d6a444ff63b14d94dead8cc10b221ec147cd817b18254dbf28f3d70f4fa5" already present on machine
  Normal  Created    31s   kubelet            Created container place-tools
  Normal  Created    30s   kubelet            Created container step-build-and-push
  Normal  Pulled     30s   kubelet            Container image "gcr.io/kaniko-project/executor@sha256:68bb272f681f691254acfbdcef00962f22efe2f0c1e287e6a837b0abe07fb94b" already present on machine
  Normal  Started    30s   kubelet            Started container place-tools
  Normal  Started    29s   kubelet            Started container step-build-and-push
  Normal  Pulled     29s   kubelet            Container image "gcr.io/tekton-releases/github.com/tektoncd/pipeline/cmd/imagedigestexporter:v0.16.2" already present on machine
  Normal  Created    29s   kubelet            Created container step-write-digest
  Normal  Started    29s   kubelet            Started container step-write-digest
  Normal  Pulled     29s   kubelet            Container image "docker.io/stedolan/jq@sha256:a61ed0bca213081b64be94c5e1b402ea58bc549f457c2682a86704dd55231e09" already present on machine
  Normal  Created    29s   kubelet            Created container step-digest-to-results
  Normal  Started    29s   kubelet            Started container step-digest-to-results
```

In this case the latency would be calculated between the time the container is scheduled, which we can get from:
```
  - lastProbeTime: null
    lastTransitionTime: "2021-04-13T11:32:16Z"
    status: "True"
    type: PodScheduled
```

and:

```
  Normal  Started    29s   kubelet            Started container step-build-and-push
```
So, probably 4s?
Looking at the pod container statuses:
```
  containerStatuses:
  - containerID: containerd://47d0cc3a9c125f9fe9da62aad08e104f327d46684cfe51c2fcc0cba3b4f812ba
    image: sha256:27d51d7fa864da161bf917fd36b8426df8fc2756959d952d2ec4703491608758
    imageID: gcr.io/kaniko-project/executor@sha256:68bb272f681f691254acfbdcef00962f22efe2f0c1e287e6a837b0abe07fb94b
    lastState: {}
    name: step-build-and-push
    ready: false
    restartCount: 0
    started: false
    state:
      terminated:
        containerID: containerd://47d0cc3a9c125f9fe9da62aad08e104f327d46684cfe51c2fcc0cba3b4f812ba
        exitCode: 0
        finishedAt: "2021-04-13T11:32:40Z"
        message: '[{"key":"StartedAt","value":"2021-04-13T11:32:22.080Z","type":"InternalTektonResult"}]'
        reason: Completed
        startedAt: "2021-04-13T11:32:20Z"
```

The `startedAt` is `"2021-04-13T11:32:20Z"`, 4s earlier which matches.

Reply:
Yes, it's meant for the first step in a `TaskRun`.  I guess your way to calculate the metrics makes more sense. I'll update this in the TEP.